# Analiza kolumny system: abilities i attributes


Import danych

In [1]:
from training.analysis_functions import get_merged_bestiaries
from collections import defaultdict
import pandas as pd
import numpy as np


paths = [
    "../../pathfinder_2e_data/pathfinder-bestiary.db",
    "../../pathfinder_2e_data/pathfinder-bestiary-2.db",
    "../../pathfinder_2e_data/pathfinder-bestiary-3.db",
]

bestiary = get_merged_bestiaries(paths)

In [2]:
bestiary.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   _id             1206 non-null   object
 1   img             1206 non-null   object
 2   items           1206 non-null   object
 3   name            1206 non-null   object
 4   system          1206 non-null   object
 5   type            1206 non-null   object
 6   flags           1206 non-null   object
 7   prototypeToken  85 non-null     object
dtypes: object(8)
memory usage: 84.8+ KB


In [3]:
bestiary.head()

_id                                       img  \
0  024PqcF8yMRBrPuq  systems/pf2e/icons/default-icons/npc.svg   
1  05E3kkjoLZVjFOeO  systems/pf2e/icons/default-icons/npc.svg   
2  05wwpHHsBlxBbdkN  systems/pf2e/icons/default-icons/npc.svg   
3  0FGz2eXm0SB04sJW  systems/pf2e/icons/default-icons/npc.svg   
4  0HjVFx8qIKDCfblg  systems/pf2e/icons/default-icons/npc.svg   

                                               items                  name  \
0  [{'_id': 'B7rzf0nBJmzg8x0y', 'img': 'systems/p...    Adult White Dragon   
1  [{'_id': 'mYNDkS27YQRyy1F0', 'img': 'systems/p...          Brontosaurus   
2  [{'_id': 'WkPeg600zGONsuJz', 'img': 'systems/p...        Giant Anaconda   
3  [{'_id': 'joBYS96mXSnZC1WB', 'img': 'systems/p...  Ancient Green Dragon   
4  [{'_id': 'xlUnQcjrxeMXu5Cd', 'img': 'systems/p...        Lantern Archon   

                                              system type  \
0  {'abilities': {'cha': {'mod': 1}, 'con': {'mod...  npc   
1  {'abilities': {'cha': {'mod': 1}, 'con': {'mod...  npc   
2  {'abilities': {'cha': {'mod': -2}, 'con': {'mo...  npc   
3  {'abilities': {'cha': {'mod': 6}, 'con': {'mod...  npc   
4  {'abilities': {'cha': {'mod': 1}, 'con': {'mod...  npc   

                                               flags prototypeToken  
0  {'core': {'sourceId': 'Compendium.pf2e.pathfin...            NaN  
1  {'core': {'sourceId': 'Compendium.pf2e.pathfin...            NaN  
2  {'core': {'sourceId': 'Compendium.pf2e.pathfin...            NaN  
3  {'core': {'sourceId': 'Compendium.pf2e.pathfin...            NaN  
4  {'core': {'sourceId': 'Compendium.pf2e.pathfin...            NaN

```unpack_column``` - Funkcja która rozpakowuje słownik i zmienia go w dataframe. Printuje pierwszą wartość tabeli i długość

In [4]:
from training.analysis_functions import unpack_column

In [5]:
bestiary_system = unpack_column(bestiary, "system")

## Abilities

In [6]:
abilities = unpack_column(bestiary_system, "abilities")

In [7]:
abilities.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cha     1206 non-null   object
 1   con     1206 non-null   object
 2   dex     1206 non-null   object
 3   int     1206 non-null   object
 4   str     1206 non-null   object
 5   wis     1206 non-null   object
dtypes: object(6)
memory usage: 66.0+ KB


In [8]:
abilities.head()

cha         con         dex          int          str         wis
0   {'mod': 1}  {'mod': 5}  {'mod': 2}   {'mod': 1}   {'mod': 7}  {'mod': 2}
1   {'mod': 1}  {'mod': 5}  {'mod': 0}  {'mod': -4}   {'mod': 9}  {'mod': 2}
2  {'mod': -2}  {'mod': 6}  {'mod': 3}  {'mod': -4}   {'mod': 7}  {'mod': 3}
3   {'mod': 6}  {'mod': 5}  {'mod': 4}   {'mod': 6}   {'mod': 7}  {'mod': 5}
4   {'mod': 1}  {'mod': 1}  {'mod': 3}  {'mod': -1}  {'mod': -5}  {'mod': 1}

Sprawdzenie jakie klucze występują w słownikach - czy tylko ```mod```?

In [9]:
for col in abilities.columns:
    print(f"{col}: ", end="")
    temp = unpack_column(abilities, col)
    print("ok" if len(temp.columns) > 0 or "mod" not in temp.columns else "More keys\n")

cha: ok
con: ok
dex: ok
int: ok
str: ok
wis: ok


Same mody, więc możnachyba zmienić te słowniki na wartość

In [10]:
abilities.head()

cha         con         dex          int          str         wis
0   {'mod': 1}  {'mod': 5}  {'mod': 2}   {'mod': 1}   {'mod': 7}  {'mod': 2}
1   {'mod': 1}  {'mod': 5}  {'mod': 0}  {'mod': -4}   {'mod': 9}  {'mod': 2}
2  {'mod': -2}  {'mod': 6}  {'mod': 3}  {'mod': -4}   {'mod': 7}  {'mod': 3}
3   {'mod': 6}  {'mod': 5}  {'mod': 4}   {'mod': 6}   {'mod': 7}  {'mod': 5}
4   {'mod': 1}  {'mod': 1}  {'mod': 3}  {'mod': -1}  {'mod': -5}  {'mod': 1}

### Zamiana słowników na wartości *mod*

In [11]:
from copy import deepcopy


abilities_backup = deepcopy(abilities)

// Znalazłam ładniejszy sposób na podmianę

In [12]:
abilities = abilities.applymap(lambda x: x.get("mod"), na_action="ignore")

In [13]:
abilities.head()

cha  con  dex  int  str  wis
0    1    5    2    1    7    2
1    1    5    0   -4    9    2
2   -2    6    3   -4    7    3
3    6    5    4    6    7    5
4    1    1    3   -1   -5    1

**abilities do uwzględnienia w predykcji**

## Analiza attributes

In [14]:
attributes = unpack_column(bestiary_system, "attributes")

In [15]:
attributes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   ac           1206 non-null   object 
 1   allSaves     1206 non-null   object 
 2   hp           1206 non-null   object 
 3   immunities   637 non-null    object 
 4   initiative   1206 non-null   object 
 5   perception   1206 non-null   object 
 6   speed        1206 non-null   object 
 7   weaknesses   463 non-null    object 
 8   resistances  344 non-null    object 
 9   hardness     10 non-null     object 
 10  adjustment   0 non-null      float64
dtypes: float64(1), object(10)
memory usage: 113.1+ KB


Brakujące wartości w kolumnach:
* immunities
* weaknesses
* resistance
W *adjustments* nic nie ma

In [16]:
attributes.head()

ac  \
0  {'details': '', 'value': 29}   
1  {'details': '', 'value': 28}   
2  {'details': '', 'value': 25}   
3  {'details': '', 'value': 41}   
4  {'details': '', 'value': 16}   

                                        allSaves  \
0  {'value': '+1 status to all saves vs. magic'}   
1                                  {'value': ''}   
2                                  {'value': ''}   
3  {'value': '+1 status to all saves vs. magic'}   
4                                  {'value': ''}   

                                                  hp  \
0  {'details': '', 'max': 215, 'temp': 0, 'value'...   
1  {'details': '', 'max': 220, 'temp': 0, 'value'...   
2  {'details': '', 'max': 175, 'temp': 0, 'value'...   
3  {'details': '', 'max': 315, 'temp': 0, 'value'...   
4  {'details': '', 'max': 20, 'temp': 0, 'value':...   

                                          immunities  \
0  [{'type': 'cold'}, {'type': 'paralyzed'}, {'ty...   
1                                                NaN   
2                                                NaN   
3  [{'type': 'paralyzed'}, {'type': 'poison'}, {'...   
4                                                NaN   

                  initiative     perception  \
0  {'ability': 'perception'}  {'value': 20}   
1  {'ability': 'perception'}  {'value': 16}   
2  {'ability': 'perception'}  {'value': 17}   
3  {'ability': 'perception'}  {'value': 30}   
4  {'ability': 'perception'}   {'value': 6}   

                                               speed  \
0  {'details': 'climb on ice only', 'otherSpeeds'...   
1                   {'otherSpeeds': [], 'value': 35}   
2  {'otherSpeeds': [{'type': 'climb', 'value': 30...   
3  {'details': 'trackless step, woodland stride',...   
4  {'otherSpeeds': [{'type': 'fly', 'value': 40}]...   

                        weaknesses                     resistances hardness  \
0  [{'type': 'fire', 'value': 10}]                             NaN      NaN   
1                              NaN                             NaN      NaN   
2                              NaN                             NaN      NaN   
3                              NaN                             NaN      NaN   
4   [{'type': 'evil', 'value': 3}]  [{'type': 'fire', 'value': 3}]      NaN   

   adjustment  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [17]:
# attributes.columns.values

Kolumny:
* ac - do sprawdzenia (przydatne)
* allSaves - do sprawdzenia
* hp - do sprawdzenia (przydatne)
* immunities - do sprawdzenia (sporo braków)
* initiative - do sprawdzenia
* perception - do sprawdzenia (wygląda obiecująco)
* weaknesses - do sprawdzenia (sporo braków)
* resistance - do sprawdzenia (sporo braków)

### Analiza attributes: ac

In [18]:
ac = unpack_column(attributes, "ac")

In [19]:
ac.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   details  1206 non-null   object
 1   value    1206 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 28.3+ KB


In [20]:
ac.head()

details  value
0             29
1             28
2             25
3             41
4             16

In [21]:
ac = ac.replace("", np.NAN)

In [22]:
print("Null values:")
ac["details"].isnull().sum()

Null values:


1147

In [23]:
ac["details"].value_counts()

all-around vision                       25
fed by metal                             6
(21 with shield raised)                  2
(25 with shield raised)                  2
+1 status vs. evil creatures             2
(22 when broken); construct armor        2
(20 with shield raised)                  2
22 with shield raised                    2
(18 with shield raised)                  1
(32 when broken); construct armor        1
(26 when broken); construct armor        1
(31 with guard raised)                   1
(26 with shield raised)                  1
(27 without alien vestment)              1
(29 with shields raised)                 1
(35 when broken); construct armor        1
(22 against prey; see Hunt Prey)         1
(10 from inside the Dimensional Pit)     1
or 36 vs. non-magical                    1
(14 when broken); construct armor        1
(13 when broken); construct armor        1
(22 with shield raised)                  1
(15 when broken); construct armor        1
(12 when br

Można olać całość i uwzglednic jedynie **value**

Przynajmniej tak mi się wydaje

**ac value do uwzględnienia w predykcji**

### Analiza attributes: allSaves

In [24]:
allSaves = unpack_column(attributes, "allSaves")

In [25]:
allSaves.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   value   1202 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


In [26]:
allSaves.head()

value
0  +1 status to all saves vs. magic
1                                  
2                                  
3  +1 status to all saves vs. magic
4

In [27]:
allSaves = allSaves.replace("", np.NAN)

In [28]:
print("Null values:")
allSaves["value"].isnull().sum()

Null values:


904

In [29]:
allSaves["value"].value_counts()

+1 status to all saves vs. magic                                                           207
+2 status to all saves vs. magic                                                             8
+1 status to all saves vs. positive                                                          8
+1 status to all saves vs. magic, +4 status to all saves vs. mental                          7
+1 to all saves vs. magic                                                                    7
circle of protection                                                                         5
+1 status to all saves vs. magic, +2 status to all saves vs. mental                          3
+4 status to all saves vs. mental or divine                                                  3
+2 circumstance to all saves vs. disease                                                     3
+2 status to all saves vs. magic, +3 status to all saves vs. divine magic                    3
+4 status to will saves vs. mental                

Nie wiem jak to uwzględnić

Gdyby to było w jakiejś postaci typu tabela kategorii i wartość dodana np. cold: +2

ale nie jest

więc chyba można olać
CR: wywalmy to

### Analiza attributes: *hp*

In [30]:
hp = unpack_column(attributes, "hp")
hp = hp.replace("", np.NAN)

In [31]:
hp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   details  195 non-null    object
 1   max      1206 non-null   int64 
 2   temp     1206 non-null   int64 
 3   value    1206 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 47.1+ KB


In [32]:
hp.head()

details  max  temp  value
0     NaN  215     0    215
1     NaN  220     0    220
2     NaN  175     0    175
3     NaN  315     0    315
4     NaN   20     0     20

In [33]:
print("Null values:")
hp["details"].isnull().sum()

Null values:


1011

In [34]:
hp["details"].value_counts()

negative healing                         67
(negative healing)                        8
troop defenses                            7
negative healing, rejuvenation            6
regeneration 20 (deactivated by evil)     6
                                         ..
repair mode                               1
fast healing 2 (while touching fire)      1
rejuvenation                              1
filth wallow                              1
death umbra                               1
Name: details, Length: 84, dtype: int64

analogicznie jak powyżej

In [35]:
hp["temp"].value_counts()

0    1206
Name: temp, dtype: int64

temp to zawsze "wartość pusta" => 0 - można olać?

**Wydaje mi się że stąd można "ukraść" max oraz value**

In [36]:
hp[hp["max"] == hp["value"]].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   details  195 non-null    object
 1   max      1206 non-null   int64 
 2   temp     1206 non-null   int64 
 3   value    1206 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 47.1+ KB


*Podsumowanie*

* details - nie będziemy uwzględniać
* max - można, czemu nie (int + brak wartości pustych) (*update max == value*)
* temp - i tak same 0 w bestiariuszach, więc jak na ten moment to raczej nie
* **value - na pewno bierzemy**

*CR: details -> Tutaj w sumie połowa potworków coś ma. Najczęstrze jest to negative healing - ale tylko ~67 wartości.
No ale chyba jest to trudne do uwzględnienia więc można usunąć*

### Analiza attributes: *initiative*

In [37]:
initiative = unpack_column(attributes, "initiative")

In [38]:
initiative.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   ability  1206 non-null   object
dtypes: object(1)
memory usage: 18.8+ KB


In [39]:
initiative.head()

ability
0  perception
1  perception
2  perception
3  perception
4  perception

In [40]:
initiative.value_counts()

ability   
perception    1205
dec              1
dtype: int64

In [41]:
initiative.loc[initiative["ability"] != "perception"]

ability
85     dec

In [42]:
bestiary.iloc[85, :]

_id                                                BRo0snV2sH6TFuh6
img                        systems/pf2e/icons/default-icons/npc.svg
items             [{'_id': '7aI3LWA9cnoKonHZ', 'img': 'systems/p...
name                                                       Glabrezu
system            {'abilities': {'cha': {'mod': 7}, 'con': {'mod...
type                                                            npc
flags             {'core': {'sourceId': 'Compendium.pf2e.pathfin...
prototypeToken                                                  NaN
Name: 85, dtype: object

In [43]:
bestiary.iloc[85, 4].get("details").get("level")

{'value': 13}

Wszystkie potworki oprócz Glabrezu mają wpisane "perception"

Też raczej można olać
*CR: olej*

### Analiza attributes: *perception*

In [44]:
perception = unpack_column(attributes, "perception")

In [45]:
perception.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   value   1206 non-null   int64
dtypes: int64(1)
memory usage: 18.8 KB


In [46]:
perception.head()

value
0     20
1     16
2     17
3     30
4      6

Dane w "przyjaznej" formie, są uwzględniane w statystykach - *mogą się przydać*

**można uwzględnić perception w predykcji**
*CR: ok*

### Analiza attributes: *speed*

In [47]:
speed = unpack_column(attributes, "speed")

In [48]:
speed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1206 entries, 0 to 364
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   details      124 non-null    object 
 1   otherSpeeds  1206 non-null   object 
 2   value        1205 non-null   float64
dtypes: float64(1), object(2)
memory usage: 37.7+ KB


In [49]:
speed.head()

details  \
0                climb on ice only   
1                              NaN   
2                              NaN   
3  trackless step, woodland stride   
4                              NaN   

                                         otherSpeeds  value  
0  [{'type': 'climb', 'value': 30}, {'type': 'fly...   30.0  
1                                                 []   35.0  
2  [{'type': 'climb', 'value': 30}, {'type': 'swi...   30.0  
3  [{'type': 'fly', 'value': 200}, {'type': 'swim...   50.0  
4                     [{'type': 'fly', 'value': 40}]    0.0

In [50]:
speed.details.isnull().sum()

1082

In [52]:
def unique_val_list(column, key):

    unique_val = defaultdict(lambda: 0)
    temp = column[column.notnull()]

    for row in temp.values:
        for d in row:
            unique_val[d.get(key)] += 1

    return unique_val

Rodzaje other speeds

In [53]:
unique_speeds = unique_val_list(speed["otherSpeeds"], "type")

In [54]:
print(len(unique_speeds))
display(unique_speeds)

8


defaultdict(<function __main__.unique_val_list.<locals>.<lambda>()>,
            {'climb': 168,
             'fly': 416,
             'swim': 230,
             'burrow': 86,
             'Ice Burrow': 1,
             'ice-burrow': 1,
             'Air Walk': 2,
             'Water Walk': 1})

Wydaje mi się, żę też to można uwzględnić tylko, że trzeba to jakoś rozbić na kolumny i zdecydować co zrobić z wartościami brakującymi - pierwszy strzał to dać tam 0 (jak ktoś nie umie latać to fly speed == 0)

**do uwzględnienia?**
*CR: Tak. Kategoria: na później*

### Analiza attributes: *immunities*

In [ ]:
attributes.immunities

In [ ]:
attributes.immunities.info()

Niemal połowa potworków (z 1206) nie ma nic wpisane w tej kolumnie

Wydaje mi się, że się raczej nie przyda

Aczkolwiek można porozbijać na kolumny - może coś da
*CR: na później (Jak osiągniemy wystarczająco dobry model bez tego to nie dodamy)*

In [ ]:
immunities_not_null = attributes[attributes["immunities"].notnull()].immunities

In [ ]:
unique_immunities = unique_val_list(immunities_not_null, "type")
print(len(unique_immunities))
display(unique_immunities)

mamy 52 unikatowe wartości - kolasyczne pytanie:


### Analiza attributes: *weaknesses*

In [197]:
attributes.weaknesses

0          [{'type': 'fire', 'value': 10}]
1                                      NaN
2                                      NaN
3                                      NaN
4           [{'type': 'evil', 'value': 3}]
                      ...                 
360                                    NaN
361    [{'type': 'cold-iron', 'value': 2}]
362                                    NaN
363                                    NaN
364                                    NaN
Name: weaknesses, Length: 1206, dtype: object

In [202]:
not_null_weakness = attributes[pd.notnull(attributes.weaknesses)].weaknesses

In [203]:
not_null_weakness.head()

0                      [{'type': 'fire', 'value': 10}]
4                       [{'type': 'evil', 'value': 3}]
6    [{'type': 'cold-iron', 'value': 5}, {'type': '...
7                       [{'type': 'good', 'value': 5}]
9    [{'type': 'positive', 'value': 5}, {'type': 's...
Name: weaknesses, dtype: object

In [234]:
print(max([len(v) for v in not_null_weakness.values]))

4


Potworki mają 0-4 słabości

In [204]:
unique_weaknesses = unique_val_list(not_null_weakness, "type")

In [205]:
display(unique_weaknesses)
print("all weaknesses", sum(unique_weaknesses.values()))
print("unique values:", len(unique_weaknesses))

defaultdict(<function __main__.unique_val_list.<locals>.<lambda>()>,
            {'fire': 80,
             'evil': 33,
             'cold-iron': 115,
             'good': 82,
             'positive': 15,
             'slashing': 24,
             'cold': 48,
             'bludgeoning': 7,
             'silver': 14,
             'chaotic': 8,
             'vampire-weaknesses': 3,
             'lawful': 12,
             'area-damage': 54,
             'splash-damage': 54,
             'vulnerable-to-sunlight': 3,
             'axe-vulnerability': 7,
             'sonic': 8,
             'salt': 3,
             'piercing': 4,
             'vorpal': 1,
             'salt-water': 2,
             'electricity': 5,
             'orichalcum': 4,
             'water': 10,
             'mental': 6,
             'negative': 1,
             'acid': 1,
             'air': 1,
             'earth': 1,
             'emotion': 1})

all weaknesses 607
unique values: 30


Większość potworków nie ma słabości, około 10% ma dwie, a więcej (do 4) zdarza się sporadycznie

*Zawsze można spróbować to uwzględnić może pomoże*
*CR: Dodałabym do kategorii na później, bo wydaje mi się to istotną mechaniką gry.*

### Analiza attributes: *resistances*

In [235]:
attributes.resistances

0                                                    NaN
1                                                    NaN
2                                                    NaN
3                                                    NaN
4                         [{'type': 'fire', 'value': 3}]
                             ...                        
360                    [{'type': 'mental', 'value': 20}]
361                                                  NaN
362    [{'exceptions': ['adamantine', 'bludgeoning'],...
363    [{'doubleVs': ['non-magical'], 'exceptions': [...
364                                                  NaN
Name: resistances, Length: 1206, dtype: object

In [236]:
not_null_resistances = attributes[pd.notnull(attributes.resistances)].resistances

In [237]:
unique_resistances = unique_val_list(not_null_resistances, "type")

In [238]:
display(unique_resistances)
print("all resistances", sum(unique_resistances.values()))
print("unique values:", len(unique_resistances))

defaultdict(<function __main__.unique_val_list.<locals>.<lambda>()>,
            {'fire': 66,
             'all-damage': 30,
             'bludgeoning': 47,
             'piercing': 70,
             'cold': 45,
             'poison': 39,
             'electricity': 34,
             'slashing': 63,
             'acid': 21,
             'physical': 81,
             'precision': 7,
             'protean-anatomy': 6,
             'mental': 35,
             'negative': 12,
             'sonic': 8,
             'force': 1,
             'critical-hits': 2,
             'evil': 1,
             'air': 1,
             'earth': 1,
             'plant': 1,
             'water': 1})

all resistances 572
unique values: 22


*To też można uwzględnić*

Potencjalnie dużo kolumn może się stwożyć - podobnie jak wyżej tam gdzie brak wartości proponuję 0
*CR: Dodałabym do kategorii na potem.*

### Analiza attributes: *hardness*

In [239]:
attributes.hardness

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
      ... 
360    NaN
361    NaN
362    NaN
363    NaN
364    NaN
Name: hardness, Length: 1206, dtype: object

In [244]:
attributes.hardness.info()

<class 'pandas.core.series.Series'>
Int64Index: 1206 entries, 0 to 364
Series name: hardness
Non-Null Count  Dtype 
--------------  ----- 
10 non-null     object
dtypes: object(1)
memory usage: 51.1+ KB


In [245]:
attributes.hardness.head()

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: hardness, dtype: object

In [247]:
hardness = unpack_column(attributes[pd.notnull(attributes.hardness)], "hardness")

In [248]:
display(hardness)

value
109      2
185      6
286      9
436     10
306     10
17      15
56       8
251     10
253     14
351      3

Nie wiem co się stało przy rozpakowywaniu, ale tu są tylko *values*

*Wydaje mi się, że to też można użyć*

Wartość specyficzna - głównie dla obiektów -> możliwe, iż powyższe zdanie jest nieaktualne

*CR: Tu idea jest taka że hardness działa tak jak resistance na wszystkie obrażenia. Czyli u nas można przepisać wartość z hardness do all-damage resistance.
Ale jest ich też strasznie mało więc można usunąć.*

*Dodałabym do kategorii na potem i jeśli kiedyś będziemy uwzględniać resistance to możemy przepisać tam wartośći.*

### Analiza attributes: *adjustment*

In [250]:
attributes.adjustment.head()

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
Name: adjustment, dtype: float64

In [251]:
attributes[attributes.adjustment.notnull()].adjustment

Series([], Name: adjustment, dtype: float64)

In [252]:
attributes.adjustment.info()

<class 'pandas.core.series.Series'>
Int64Index: 1206 entries, 0 to 364
Series name: adjustment
Non-Null Count  Dtype  
--------------  -----  
0 non-null      float64
dtypes: float64(1)
memory usage: 51.1 KB


Tutaj są same nulle, więc można usunąć

## Podsumowując:
* Abilities - wszystkie wartości do uwzględnienia
* Attributes:
    * ac - value do uwzględnienia
    * allSaves - raczej nie
    * hp - value do uwzględnienia + ewentualnie max
    * initiative - raczej nie jest potrzebne
    * perception - można uwzględnić
    * speed - też można spróbować uwzględnić
    * immunities - jak na razie nie, ale można spróbować
    * weaknesses - do tego nie jestem w 100% przekonana, ale zawsze można dodać
    * resistances - analogicznie jak wyżej
    * hardness - myślę, że można uwzględnić
    * adjustment - tutaj nic nie ma (dosłownie)